Round 2: CI/CD deployment guide

yorinasub17 opened the pull request:
Since the other PR (https://github.com/gruntwork-io/gruntwork-io.github.io/pull/265) was getting big and slow, I decided to open a new PR for round 2 feedback.

Link to rendered view: https://github.com/gruntwork-io/gruntwork-io.github.io/blob/yori-cicd-guide/_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+With all the core concepts out of the way, let's now discuss how to configure a production-grade CI/CD workflow for
+infrastructure code, using a platform that looks something like this:
+
+.Architecture of platform for running Terraform/Terragrunt CI/CD workflows.
```
I found this diagram to be a bit confusing.

First, in the CI Servers box, the CI icons on the left side in the corner and then multiple raised boxes didn't make sense to me. Perhaps just have one raised box labeled "CI server" and leave off the icons?

Next, why are there multiple Lambda functions, but an arrow to just one of them? Also, the "Invoke Lambda Functions" text is too small to read. 

In the ECS box, I'd leave off the "ECS Fargate" icon, then rotate the "ECS Fargate" text 90deg and increase font size.

bwhaley replied:
I actually wonder if CloudCraft is the best tool here. A flat 2D diagram might be more effective.

yorinasub17 replied:
Ok I updated the image given your feedback: 93034c5

What do you think?

> Perhaps just have one raised box labeled "CI server" and leave off the icons?
IMO, the icons are important because they give a visual cue as to what this CI server should be.

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+relying on a single platform for implementing the entire workflow. Instead, we recommend a hybrid solution that takes
+advantage of the strengths of each platform, and cover the weaknesses. The design looks as follows:
+
+- Use any generic CI/CD server you wish to integrate with your VCS repositories.
```
"This could be Jenkins, CircleCI, GitLab, or whatever you prefer."

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+Gruntwork ECS Deploy Runner Stack::
+  This is a stack you can deploy in your AWS account that sets up an ECS task with a customizable docker container for
+  running `terraform validate`, `terraform plan`, and `terraform apply`, or the Terragrunt equivalent. It is also
+  extensible to support other commands as well, such as running `go test` for Terratest or `packer build` for building
```
Perhaps make it clear that it can be extended to run *any* arbitrary command!

yorinasub17 replied:
I think I will keep as is, since that could be seen as a security risk if someone glosses over it.

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+
+
+[[options_for_deploy_server]]
+=== Options for deploy server
```
Observation: This section directly contrasts Gruntwork ECS Deploy Runner Stack with Terraform Enterprise, and paints Gruntwork's solution in a better light. Them are fightin' words.

yorinasub17 replied:
That was not the intention, so highlighted the strengths of TFE better: 0093729

What do you think of the updated wording?

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+
+Run your infrastructure deployment workloads in a https://aws.amazon.com/vpc/[Virtual Private Cloud (VPC)] to isolate
+the workloads in a restricted network topology (see link:/guides/networking/how-to-deploy-production-grade-vpc-aws[How
+to deploy a production-grade VPC on AWS] for more information on VPCs). Configure to run all workloads in private
```
```suggestion
to deploy a production-grade VPC on AWS] for more information on VPCs). Configure it to run all workloads in private
```

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+each pipeline, you can reduce the blast radius of the damage that can be done with each set of credentials. At a minimum,
+you should have two versions of the infrastructure deployment system: one for deploying the application code, and which
+only has the minimal permissions necessary for deploying that application; and one for deploying infrastructure code,
+which has full access to the environments.
```
```suggestion
which has more access to the environments.
```

Even for the infra deployments, it would be ideal if it could have reduced permissions. In fact, to be compliant with CIS, you couldn't have a deployment system with full access.

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+[[use_approval_flows]]
+=== Use approval flows
+
+It is important that human review is baked into each deployment. As covered in <<cicd_for_infrastructure_code>>, it is
```
`<<cicd_for_infrastructure_code>>` <== does this need to be fixed?

yorinasub17 replied:
Ah yes. Good catch!

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
This list might be more readable on separate lines.

yorinasub17 replied:
Agreed, but due to the limitations of asciidoc definition lists it's a bit hard to massage this. I'm going to keep it as is for now and we can revisit later.

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
As written, the last sentence here is a confusing.

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```suggestion
  secrets, for example. The IAM policies are set up such that the IAM user for CI only has
```

bwhaley replied:
I don't understand what github's suggested change is doing here. All I'm suggesting is to drop the phrase "It is important to note that" from the beginning of that sentence.

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
Consider adding a blank line above this.

bwhaley commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
Should we use a newer release?

yorinasub17 replied:
This is actually the latest:

<img width="874" alt="Configuring_CircleCI_-_CircleCI_-_Mozilla_Firefox" src="https://user-images.githubusercontent.com/430092/76347521-067a1e80-62c4-11ea-9e7d-cdf4736e0a48.png">

josh-padnick commented:
>  It's a huge doc - it took me almost 2 hours to review just now, and it's my second pass. 

I've continued to snooze my reminder to review this but am fearful to engage for this very reason.

> My only overarching concern is that this is a LOT of information. 

Although I haven't read the doc comprehensively, I've browsed it, and it occurred to me that perhaps the Production Deployment Guide format breaks down after a certain volume of content. After all, including code samples, this clocks in at over 16,000 words! Assuming 250 words per page, that would make this a 64-page book, or about 1/4 of a typical O'Reilly book. And we're ultimately rendering this all on a single page, too.

I've been thinking lately that we probably need a first-class way to handle deployment guides that treats all this content like documentation you might see on https://stripe.com/docs. Rather than a "guide," we'd consider a CI/CD pipeline as a "use case" or "core concept" and include a collection of docs specially on that concept. Another take on this might be that this is how we build up the mythic DevOps Blog Post series, one "guide" or "use case" or "core concept" at a time. 

Anyway, I'm driving this discussion in places that go beyond the scope of this PR so we should address that elsewhere. For now, we should finish out the guide as is and get it live so we can start getting value out of it.

'cc @brikis98 @tsmit291 

P.S. Reminder that this repo (and therefore this discussion) is public.

brikis98 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+
+It is a good practice to define and store the deployment pipeline as code in the same repo that it is used. For example,
+you should define the CI/CD deployment pipeline for your infrastructure code in the `modules` and `live` repositories.
+However, this means that anyone with access to those repositories are free to modify the pipeline, __even on feature
```
```suggestion
However, this means that anyone with access to those repositories could modify the pipeline, __even on feature
```

brikis98 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
I'd say limiting the CI server to read-only access is prob good enough

yorinasub17 replied:
Updated: d5227b3

brikis98 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```suggestion
However, the reality is that __any__ CI/CD solution can likely be compromised if an attacker has full access to your source code.
```

brikis98 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
This part here and the threat model section above goes into a LOT of detail on the security aspects. This is important, but it's also quite long...

Here's an idea: what if we extract the threat model into a separate doc? We could, for example, put it in a Markdown doc in the `module-security` repo and link to that doc from this guide by using the `/infrastructure-as-code-library` page on gruntwork.io. Then, in this guide, we can reduce the security details to a much smaller subset: namely, that deploying arbitrary infra changes requires arbitrary (admin) permissions, and giving a CI server admin permissions is a bad idea as (a) CI servers aren't known to be super secure, (b) you give your whole team access to those servers and let them execute arbitrary code on them, so now everyone on your team is effectively an admin. Our solution mitigates this by giving the admin permissions solely to a separate deploy server. For more details on the threat model, see `<this doc>`.

yorinasub17 replied:
I know we want to pare down the docs, but I think I am going to keep it as is for now. My main rationale for this is that the threat model and the security aspects are precisely why this isn't as simple as "use Atlantis." So having the doc focus on it and spend a lot of time on it is important IMO.

brikis98 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
Do you need all these snippets and examples? Or could you just go straight into the next section about deploying the ECS deploy runner?

yorinasub17 replied:
I was just following what you had in the k8s guide. IMO, I think it is good to reiterate at least the folder structure so we know what assumptions we are working with.

With that said, the actual code contents are probably not necessary so removed: d7775d3

brikis98 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
Neat 👍

brikis98 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
I think having these all defined within the guide is good for now, but when we switch to the service catalog model, and are no longer setting up CI / CD for customers, we'll want to create reusable scripts and copy/pasteable code for customers to use to make the setup process easier.

yorinasub17 replied:
Yup makes sense, although I think more thought is needed in the script design. E.g can we make a drop in file setup?

brikis98 replied:
Def worth thinking through!

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+<<production_grade_design>>::
+  An overview of how to configure a secure, scalable, and robust CI/CD workflow that you can rely on for your
+  production application and infrastructure code. To get a sense of what production-grade means, check out
+  link:/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library#production_grade_infra_checklist[The production-grade infrastructure checklist].
```
This will actually work once it deploys, as the URLs are constructed with jekyll.

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+.Components of the International Space Station
+image::/assets/img/guides/infrastructure-cicd-pipeline/iss-components.png[]
+
+The ISS consists of dozens of components, as shown in the image above. Each component is built by a team from a
```
I removed the caption as it doesn't add too much value given the figure title in the image itself.

yorinasub17 replied:
6a1ff2a32079bc29ea5079308d95d44c517885b9

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+.Many teams employ a practice of working on their features over long periods of time on isolated branches. These long lived feature branches have higher chances of having big conflicts to resolve when integrated.
+image::/assets/img/guides/infrastructure-cicd-pipeline/feature-branch-merge-conflict.png[]
+
+In contrast, the Continuous Integration and Continuous Delivery model of development promotes more cross team
```
In the asciidoc rendered view (on the website) it will look ok. See how the caption is rendered here:

<img width="787" alt="How_to_deploy_a_production-grade_Kubernetes_cluster_on_AWS_-_Mozilla_Firefox" src="https://user-images.githubusercontent.com/430092/76339997-17bd2e00-62b8-11ea-97b2-1d8aa4aec5ea.png">

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+  environment as a series of configuration files for the modules that specify the various parameters specific to
+  development (e.g sizing small instances, naming them with `dev`, using accounts that are accessible by all developers,
+  etc). If the modules are "blueprints" then the live configuration contain the "houses" that were built using the
+  "blueprints."
```
Good idea!

yorinasub17 replied:
0fde26a0a5ed26c84899e2bd61752416d5c5a791

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+`infrastructure-live` for live configuration). Organizing your infrastructure code in this way ensures that you can
+better test the code before they are used to deploy your environments.
+
+Isolating modules from live configuration allows you to:
```
Ok. I guess it is fairly clear that these properties come out of the code organization without much explanation.

yorinasub17 replied:
4995cd6f54fce35fd1231b68aa02d21dab059162

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+The following covers the steps of a typical CI/CD workflow. Any CI/CD workflow, whether it be for infrastructure code or
+application code, generally follow these steps. However, the details of some of the steps differ widely. We'll go over
+each step of the process and then highlight the differences in implementation for the three flavors of code we covered:
+application code, infrastructure modules, and live infrastructure configuration.
```
Roger: 61a8a1b

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+push code back to the central repository without breaking the main line of code (trunk) that everyone else is working
+on.
+
+It is important to keep the trunk clean in a trunk-based development model, as all new work starts from the trunk.
```
I didn't think so at first, and then double checked the trunk-based development section and indeed you are right that we cover each of these already. The perils of working on the same document over multiple weeks.

Removed: 6fbc32d

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+  Locally testing live infrastructure config is more difficult than either application code or infrastructure modules.
+  Unlike with infrastructure modules, it is difficult to deploy the live infrastructure config temporarily as the code
+  is tied to a specific live environment by nature of the code. After all, this is the configuration to manage live
+  infrastructure. To illustrate this point, consider a scenario where you are working on updating the cross account IAM
```
Done: f50948a

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+  infrastructure. To illustrate this point, consider a scenario where you are working on updating the cross account IAM
+  roles to access your environments, and you are at the point of reflecting your changes to prod. Would you want to
+  deploy that code to your live production environment off of an unreviewed branch? +
+  The only real test you can do for live infrastructure config is to do a dry run of your infrastructure code. Most Infrastructure as Code tools support a dry run of the code to check what it would do against your environment. For example, with Terraform, you could run `terraform plan` to sanity check the planned actions Terraform will take. This is especially useful for sanity checking a fresh c ...
```
I tried a few ways, but in the end could not figure out a way to do this in the definition list. I think I need to switch from definition lists, but then it makes it hard to do the switch over from side-by-side-comp => overall explanation at the end.

I think I am going to keep as is for now, and then revisit later, since I am already overbudget on this project and guide.

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+valuable insights on the code that you might not have thought of.
+
+How thorough the review should be depends on the amount of automated testing. If you have a comprehensive test suite
+that everyone on the team has confidence in, then it is not necessary to have a thorough review process. On the other
```
> These are hard things to test for that a thorough and well executed code review will highlight.

I was trying to get at the fact that infrastructure code requires an even more thorough review to look for and account for bugs because you don't have automated testing, but you are right that I probably went too overboard in the claim here.

Removed as Jim suggested to avoid the politics: 9e42e51

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+
+To help with code review, you should also set up a CI server (such as Jenkins or CircleCI) with commit hooks that
+automatically trigger testing of any branch that is submitted for review. Running the automated tests in this fashion
+not only ensures that the code passes all the tests, but also verifies that it runs on multiple platforms and not just
```
Reworded: f985644

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+  for live infrastructure config because in practice you will end up deploying every commit of trunk.
+
+
+It is worth expanding a bit on the reason why live infrastructure config does not have any release artifact. To
```
I want to keep this section. This was a major source of confusion for Ben, and I suspect for a large number of people as well.

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+==== Deploy
+
+Now that you have a release artifact, the final stage of the process is to deploy the code. The act of deploying the
+code immediately after generating the release is Continuous Deployment (CD).
```
Ah I didn't know that distinction. I am also not sure a lot of people know that distinction either. Do you think it is worth describing that in this document somewhere?

In any case, I am going to remove that sentence to avoid the accuracy problem and avoid definitions of terms, since it is really just fluff.

brikis98 replied:
It's from the book *Continuous Delivery**, IIRC. No need to define it here, just want to be sure we use the right terms.

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+  config. However, if the module is being deployed for the first time, then this will require creating a new
+  configuration in your live infrastructure config to deploy the module. In either case, the only way to deploy
+  infrastructure modules is by making the corresponding edits to the live infrastructure config to roll out the changes
+  across your environments. Note that because infrastructure changes are higher risk than application code, you will
```
Makes sense. Removed: aa4c2ec

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+Regardless of the code, the important thing about the deployment process is that you should automate as many steps of
+the process as possible. Just like with automated testing on code review, you should set up a CI server to automatically
+deploy your code across the various environments, to the extent that it is safe. You can automate many steps of the
+deployment depending on the risk profile of the change:
```
Ok. Done: ebf3ce7

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+  For application code, you should be able to automate the entire deployment as the surface area of each change should
+  be fairly small and localized to just the application. Additionally, application code have standard rollback
+  practices that make it easy to undo. Automating the deployment of application code requires automatically patching
+  your live infrastructure config. In most cases, the changes are simple enough that a single `sed` call should
```
This section was consolidated and removed: ebf3ce7

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+  world where a single name change could cause the database to be recreated, it pays off to be extra cautious with the
+  deployment. At most, you should only automatically deploy infrastructure modules to a sandbox or dev environment,
+  where the damage from a catastrophic failure causing the environment to be taken down is limited. Similar to
+  application code, automating this process could be handled with a single `sed` call that updates the ref tag in the
```
This section was consolidated and removed: ebf3ce7

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+  handled as part of the automated deployment of the live infrastructure config.
+
+Infrastructure Modules::
+  With infrastructure modules, depending on which module is being updated, an automated deployment may be risky. In a
```
Ah that is a good point. Updated the wording: e4e9e60

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+.Permissions needed for CI/CD of each code type.
+[cols="1h,2,2,2"]
+|===
+|Permissions |Application Code |Infrastructure Modules |Live Infrastructure Config
```
Removed: ab4b6e7

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+environments, whereas application code can be deployed with limited permissions to the environments. This difference is
+due to the fact that infrastructure modules and live infrastructure config can be used to manage any infrastructure
+deployed in your environments, whereas application code is scoped to infrastructure that is relevant to the application.
+Given that, the threat against CI/CD in infrastructure code is orders of magnitude higher than with application code.
```
Removed this whole section and made it more generic: ab4b6e7

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+Given that, the threat against CI/CD in infrastructure code is orders of magnitude higher than with application code.
+
+Application Code::
+  Compromising the credentials used for application code may be used to take down your application in a denial of service
```
Removed: ab4b6e7

yorinasub17 replied:
I had a naive assumption where the credentials only give you access to deploy a new version of the app, but then I realized that you can deploy a new app with the same credentials as the original app so you can get access to the app data.

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+    * Full access over all resources to cause damage to the business (e.g ability to delete the database and all its
+      backups in prod).
+
+- Attackers can originate from both external and internal sources (in relation to the organization).
```
Agreed! But since this is already long, I'll regrettably keep it as is for now instead of honing the point further.

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+
+
+[[cicd_platforms]]
+=== CI/CD platforms
```
This is painful for me since I really like the text overview. The table is very terse and it leaves a lot of room for questioning since there is no explanation. But... I guess you are right about shortening the guide and see how people receive this.

Consolidated in: 6c6d2df

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+  support for various application based workflows like parsing test results into a nice report in the UI, reacting to
+  version control events, and official docker containers for various programming languages and frameworks. However, it
+  does not have first class support for features that are useful with infrastructure code, such as encrypting logs
+  and fine grained permissions. The strength of CircleCI lies in the simplicity of its infrastructure. Being a fully
```
This section was removed and consolidated to just the table in the interest of shortening the document. 6c6d2df

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+|SaaS
+|SaaS or Self-hosted
+|Self-hosted
+|SaaS or Self-hosted
```
Updated table with your info and added in buildkite: 6c6d2df

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+relying on a single platform for implementing the entire workflow. Instead, we recommend a hybrid solution that takes
+advantage of the strengths of each platform, and cover the weaknesses. The design looks as follows:
+
+- Use any generic CI/CD server you wish to integrate with your VCS repositories.
```
Took Jim's suggestion and applied it: 73cbdea

brikis98 replied:
Whoops, had a typo in my description that I think got copied into the doc:

> that can be used to define pre-defined commands

Should be:

> that can be used to run pre-defined commands

yorinasub17 replied:
Fixed https://github.com/gruntwork-io/gruntwork-io.github.io/pull/273

yorinasub17 commented on _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc:
```diff
@@ -0,0 +1,2334 @@
+  which exposes a limited set of options and additional checks for source repository. It relies on serverless
+  technologies to limit the amount of overhead required for maintaining the system.
+
+Terraform Enterprise::
```
TFE has the sentinel policies though, which I think would be super powerful in CI/CD.

yorinasub17 replied:
Updated the wording to clarify this point: 0093729

What do you think?

brikis98 replied:
I think that clarifies some things, but doesn't quite touch on the fact that if you're allowing devs to click around TFE, provide various input variables in the UI, and kick off deployments that leverage TFE's admin permissions, some of the security benefits discussed in this doc don't quite work.

yorinasub17 replied:
I feel like using TFE as a deploy server in our workflow is orthogonal to using TFE for self-service. For example, you could use TFE just for the sentinel purpose when driving deployments in CI/CD fashion and close off the UI. Sure that lowers the value of TFE, but it doesn't make it less secure than the ECS deploy runner. And you get sentinel policies with it.

So there are extra steps one needs to take to use TFE as a deploy server (this is why being super detailed about the threat model here is important!), but it doesn't make TFE completely incompatible. Compare this with Atlantis, where you can't manually trigger deployments, or TFC where it is hosted and thus you have to share your credentials with the third party. There is no way to use those tools in the workflow we described while mitigating the threats, and hence are completely incompatible.

yorinasub17 commented:
Thanks for the detailed review all! I believe I addressed all the feedback, so at this point I am going to pull the trigger and merge this.

As a reminder, I don't have the ability to deploy this so need some help with that.

brikis98 commented:
@eak12913 Could you deploy the static website when you get a chance?

eak12913 commented:
@brikis98: Deployed!
